dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/datafile/FileServiceIT.java

```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void deleteFile_forDraftDataset() {
+        // given
+        Dataset dataset = datasetDao.find(DRAFT_DATASET_WITH_FILES_ID);
+
+        Tuple2<VersionState, Long> versionDataBefore = getLatestVersionData(dataset);
+
+        List<DataFile> files = dataset.getFiles();
+        DataFile fileToDelete = files.get(0);
+
+        // when
+        Dataset updatedDataset = fileService.deleteFile(fileToDelete.getFileMetadata());
+
+        // then
+        List<DataFile> updatedFiles = updatedDataset.getFiles();
+
+        assertThat("State and version after delete should match",
+                getLatestVersionData(updatedDataset), equalTo(versionDataBefore));
+        assertThat("File list in updated draft should not contain deleted file",
+                fileToDelete, not(in(updatedFiles))); // DataFile#equals(…) is based only on file's id
+    }
+
+    @Test
+    public void deleteFile_forReleasedDataset() {
+        // given
+        publishDataset(DRAFT_DATASET_WITH_FILES_ID);
+        Dataset dataset = datasetDao.find(DRAFT_DATASET_WITH_FILES_ID);
+
+        Tuple2<VersionState, Long> versionDataBefore = getLatestVersionData(dataset);
+
+        DatasetVersion versionBefore = dataset.getLatestVersion();
+        List<FileMetadata> filesMetadataBefore = versionBefore.getFileMetadatas();
+        FileMetadata fileToDeleteMetadata = filesMetadataBefore.get(0);
+
+        // when
+        Dataset updatedDataset = fileService.deleteFile(fileToDeleteMetadata);
+
+        // then
+        DatasetVersion versionAfter = updatedDataset.getLatestVersion();
+        Tuple2<VersionState, Long> versionDataAfter = getLatestVersionData(updatedDataset);
+
+        assertThat("State before and after delete should be different and state after delete should be DRAFT",
+                versionDataAfter._1, allOf(not(equalTo(versionDataBefore._1)), equalTo(VersionState.DRAFT)));
+        assertThat("Files' metadata in new version should not contain metadata of deleted file",
+                fileToDeleteMetadata, not(in(versionAfter.getFileMetadatas())));
+        assertThat(
+                "Deleted file should be present in dataset' file collection, as it is used by previously published version(s)",
+                fileToDeleteMetadata, in(extractFileListMetadata(updatedDataset))
+        );
+    }
+
+    @Test
+    public void deleteFiles_forDraft() {
+        // given
+        Dataset dataset = datasetDao.find(DRAFT_DATASET_WITH_FILES_ID);
+
+        Tuple2<VersionState, Long> versionDataBefore = getLatestVersionData(dataset);
+
+        List<DataFile> files = dataset.getFiles();
+        Set<FileMetadata> filesToDeleteMetadata = files.stream()
+                .map(DataFile::getFileMetadata)
+                .collect(Collectors.toSet());
+
+        // when
+        Set<Dataset> results = fileService.deleteFiles(filesToDeleteMetadata);
+
+        // then
+        Dataset updatedDataset = findNewestResult(results);
```

Comment:
Why executing `findNewestResult`? We requested delete of files from single dataset, so we should expect that in result we get set with single element inside.

Reply:
We agree on that?
```
assertThat(results, hasSize(1));
Dataset updatedDataset = results.iterator().next();
```